text/0000-collection-view-redesign.md

```diff
@@ -0,0 +1,803 @@
+  the initial slice number, e.g. 20) overhead on evaluation.  The current
+  view implementations can reduce this overhead.   While the overhead can be
+  significant for some operations, there are two things we've found:
+
+1. The overhead is less than the cost of attempting to special-case slice
+    within the library.
+2. The overhead, for some particularly offensive scenarios, is on the order of
+   30% over a custom solution.   We think this represents a minor drawback, and
+   something that could be recovered via some specific "slice view" mechanism
+   that is outside the scope of this proposal.
+
+Additionally, slicing is only an issue when it is a "drop" slice, i.e. when you are removing a known quantity from the beginning of the collection.   For `dropWhile`, `takeWhile`, `take`, `slice(0,n)` performance is on par or an improvement.
+
+There is also a question of random access.   With the current proposal, implementations for methods like `head`, `apply(n)`, `update(n)` will need to execute a fold.   For a small class of use cases, this is an issue.  For example:
+
+```scala
+val reallyBigMatrix: Array[Int] = ...
+val row1 = realylBigMatrix.view.slice(col, row*col)
+// Computations using row1
+```
+
+We feel these scenarios deserve a *different* solution than existing views, because:
+
+1. Any structure altering operations removes the performance benefit, e.g. `flatMap`, `filter`, `dropWhile`
+2. The implementation of mutable updates suffers from all the issues described before, generally failures in the presence of any structure-altering methods. 
+
+## Alternatives
+
+There were a few alternatives attempted before the current proposal:
+
+1. Continue fixing the current view code base.  This is still an option,
+   and remains on the table.
+2. Leveraging macros and syntax tree lifting to stage an AST which would
+   be interpreted (at runtime or compile-time) into a more efficient
+   operation later.   While this is still possible, we feel adding a
+   transducer + alternative view implementation does not prevent such a
+   solution from also being viable.  Additionally this proposal represents,
+   what we feel, is a much easier to maintain solution.
+
+   Note that if it is done in macros - it is pretty limited and cannot go through method dispatch borders. eg
+     ```scala
+     def foo = col.view.map(_+1)
+     def bar = foo.map(_+1).sum_!
+    ```
+    In this example macros cannot do much optimization.
+    And this is an example that shows that implementing optimizations as macros rewards users for not modularizing their code.
+    Unlike this, general optimizers could. Currently the proposed views infrastructure is quite simple and general inliner(either optimistic as Scala.js linker is, or pessimistic, as the linker that @DarkDimius'es Dotty linker is) should be able to optimize this code to a simple while cycle merging the two .map(_+1) operations together after several rounds of inlining.
+
```

Comment:
Ignore me -- it looks like this is a bug in this Github rendering. When I clicked the "View" button it looks fine.